### PR TITLE
Replace x/net websocket with gorilla websocket

### PR DESCRIPTION
**Pull request by fosterseth**
golang.org/x/net is missing features or is deprecated, so replacing with a more active library, github.com/gorilla/websocket

issue https://github.com/ghjm/sockceptor/issues/5

### CODE REVIEW

File: pkg/backends/websockets.go [hunk 1]
```diff
@@ -29,13 +29,9 @@ type WebsocketDialer struct {
...
 // NewWebsocketDialer instantiates a new WebsocketDialer backend
 func NewWebsocketDialer(address string, tlscfg *tls.Config, extraHeader string, redial bool) (*WebsocketDialer, error) {
-	addrURL, err := url.Parse(address)
-	if err != nil {
-		return nil, err
-	}
 	wd := WebsocketDialer{
 		address:     address,
-		origin:      fmt.Sprintf("http://%s", addrURL.Host),
+		origin:      fmt.Sprintf("http://%s", address),
```
Comment (ghjm) [inline pkg/backends/websockets.go]: I think the `address` parameter supplied to this function is expected to be a URL.  So you probably still have to parse out the host part to make an origin.  Maybe the scheme also needs to be http or https based on whether `address`'s URL scheme is `ws` or `wss`.

File: pkg/backends/websockets.go [hunk 2]
```diff
@@ -47,21 +43,19 @@ func NewWebsocketDialer(address string, tlscfg *tls.Config, extraHeader string,
...
-			if err != nil {
-				errf(fmt.Errorf("error creating websocket config"), true)
-				return
-			}
+			dialer := websocket.DefaultDialer
 			if b.tlscfg != nil {
-				cfg.TlsConfig = b.tlscfg
+				dialer.TLSClientConfig  = b.tlscfg
 			}
+			header := make(http.Header, 0)
```
Comment (ghjm) [inline pkg/backends/websockets.go:54]: This declaration is going to be shadowed by the `:=` a few lines below, inside the `if {}` block.  This is a go-ism - every curly brace block implies a lexical scope, so the second `:=` declares a new variable that only exists within the `if {}`.  I think you get the behavior you want by just deleting the second declaration (line 60), which will make the `header.Add` apply to the outer variable.

File: pkg/backends/websockets.go [hunk 3]
```diff
@@ -47,21 +43,19 @@ func NewWebsocketDialer(address string, tlscfg *tls.Config, extraHeader string,
...
 			if b.extraHeader != "" {
 				extraHeaderParts := strings.SplitN(b.extraHeader, ":", 2)
 				header := make(http.Header, 0)
 				header.Add(http.CanonicalHeaderKey(extraHeaderParts[0]), extraHeaderParts[1])
-				cfg.Header = header
 			}
-			conn, err := websocket.DialConfig(cfg)
+			header.Add(http.CanonicalHeaderKey("origin"), b.origin)
+			u := url.URL{Scheme: "ws", Host: b.address, Path: "/"}
```
Comment (ghjm) [inline pkg/backends/websockets.go]: Callers need to be able to specify `ws://` or `wss://` - we can't assume one or the other here.  That's why `address` needs to be a full URL.

File: pkg/backends/websockets.go [hunk 4]
```diff
@@ -136,28 +132,25 @@ func newWebsocketSession(conn *websocket.Conn) *WebsocketSession {
 // Send sends data over the session
 func (ns *WebsocketSession) Send(data []byte) error {
 	buf := ns.framer.SendData(data)
-	n, err := ns.conn.Write(buf)
+	err := ns.conn.WriteMessage(websocket.BinaryMessage, buf)
```
Comment (ghjm) [inline pkg/backends/websockets.go]: Does `WriteMessage()` take a whole message, which will then be presented in its entirety on the other side via `ReadMessage()`?  If so then we don't actually need `framer` at all here.

**Review by ghjm — CHANGES_REQUESTED** [review]

File: pkg/backends/websockets.go (hunk 2 above)
Comment (fosterseth) [reply pkg/backends/websockets.go:54]: fixed

File: pkg/backends/websockets.go (hunk 4 above)
Comment (fosterseth) [reply pkg/backends/websockets.go]: yes it takes whole message, so I removed framer.

**Comment by ghjm** [comment]
As part of this PR, we should remove the TODOs at the top of `pkg/backends/websockets.go`

**Comment by ghjm** [comment]
I was thinking that in the case where the user does `receptor --node id=foo --ws-peer address=wss://bar.com` it will fail because when it gets to `NewWebsocketDialer`, tlscfg will be nil, so it will be trying to run an unencrypted connection to a `wss://` endpoint.  It would be more user friendly to use default TLS settings in this case.

We have a client TLS configuration named `default` that is always created, and is intended for situations like this.  What I'm thinking is we check in `WebSocketDialerCfg.Run` if `cfg.TLS==""` and the scheme is `wss://`, then when we call `GetClientTLSConfig`, we pass in `default` as the config name instead of `cfg.TLS`.

File: pkg/backends/websockets.go [hunk 5]
```diff
@@ -3,19 +3,17 @@ package backends
...
 	"net/http"
 	"net/url"
 	"strings"
 	"time"
 )
 
-//TODO: Switch to better supported websockets library
-//TODO: TLS Server
-//TODO: HTTP/HTTPS proxy support
+//TODO:
```
Comment (ghjm) [inline pkg/backends/websockets.go]: No reason to leave this here

Comment (fosterseth) [reply pkg/backends/websockets.go]: fixed

**Review by ghjm — APPROVED** [review]

**Pull request merged**